### PR TITLE
Handle cancellation in places that could block

### PR DESCRIPTION
**Pull request by richardpark-msft**
We have a few unbuffered channels that in normal operation we expect to cycle rather quickly. However, since they _can_ block indefinitely (unlikely, but possible) we should allow for cancellation to occur.

Most of the complications for cancel come from leaving ourselves in an inconsistent state but the spots I've added cancellation shouldn't be problematic, as they're all before network IO has occurred.

### CODE REVIEW

File: conn.go [hunk 1]
```diff
@@ -858,6 +859,15 @@ func (c *Conn) sendFrame(ctx context.Context, fr frames.Frame, sent chan error)
 	select {
 	case c.txFrame <- frameEnvelope{Ctx: ctx, Frame: fr, Sent: sent}:
 		debug.Log(2, "TX (Conn %p): mux frame to connWriter: %s", c, fr)
+	// cancelling here should be considered safe as nothing has hit the network yet
```
Comment (jhendrixMSFT) [inline conn.go:862]: This isn't true in some cases.  Client-side close and sending of disposition frames are a few examples.  If we allow cancelling here, and it fails, then we'd need to kill the session at minimum (maybe Conn, haven't thought about this in a while).

File: session.go [hunk 2]
```diff
@@ -645,12 +645,21 @@ func (s *Session) mux(remoteBegin *frames.PerformBegin) {
...
 			s.txFrame(env.Ctx, fr, env.Sent)
+
 			if sendErr := <-env.Sent; sendErr != nil {
 				s.doneErr = sendErr
 
 				// put the error back as our sender will read from this channel
 				env.Sent <- sendErr
-				return
+
+				if errors.Is(sendErr, context.Canceled) || errors.Is(sendErr, context.DeadlineExceeded) {
```
Comment (jhendrixMSFT) [inline session.go:655]: This shouldn't be necessary.  Receiving a `*ConnError` means that the conn has died, so on the next iteration we hit this block.
```go
// conn has completed, exit
case <-s.conn.done:
	s.doneErr = s.conn.doneErr
	return
```

**Review by jhendrixMSFT — CHANGES_REQUESTED** [review]
We should discuss further.

File: conn.go (hunk 1 above)
Comment (richardpark-msft) [reply conn.go:862]: Some of the obvious problems:

* Disposition deletes handles and assumes it's either sent or the entire session dies.

**Pull request closed without merging**